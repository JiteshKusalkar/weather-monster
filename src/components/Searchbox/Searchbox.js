import styled from 'styled-components';
import React, { useState } from 'react';
import Autosuggest from 'react-autosuggest';

import data from '../../data/weather.json';
import { escape } from '../../utils/common.js';

const Searchbox = ({ id, onSuggestionSelect }) => {
  const cities = data.map(cityData => ({
    name: cityData.city.name,
    id: cityData.city.id
  }));

  const [value, setValue] = useState('');
  const [suggestions, setSuggestions] = useState(cities);

  const getSuggestions = value => {
    const inputValue = value.trim().toLowerCase();
    const inputLength = inputValue.length;

    return inputLength === 0
      ? cities
      : cities.filter(city => city.name.toLowerCase().includes(inputValue));
  };

  // To empty the input string on suggestion click
  const getSuggestionValue = suggestion => '';

  const renderSuggestion = suggestion => {
    let suggestionLabel = suggestion.name;
    const queryRegex = new RegExp(escape(value), 'ig');
    suggestionLabel = suggestionLabel.replace(queryRegex, '<b>$&</b>');

    return <Suggestion dangerouslySetInnerHTML={{ __html: suggestionLabel }} />;
  };

  const onChange = (event, { newValue }) => {
    setValue(newValue);
  };

  const onSuggestionsFetchRequested = ({ value }) => {
    setSuggestions(getSuggestions(value));
  };

  const onSuggestionsClearRequested = () => {
    setSuggestions(cities);
  };

  const onSuggestionSelected = (event, { suggestion }) => {
    onSuggestionSelect(suggestion);
  };

  const inputProps = {
    placeholder: 'Type the name of a city',
    value,
    onChange
  };

  return (
    <Wrapper>
      <Autosuggest
        id={id}
        suggestions={suggestions}
        onSuggestionsFetchRequested={onSuggestionsFetchRequested}
        onSuggestionsClearRequested={onSuggestionsClearRequested}
        getSuggestionValue={getSuggestionValue}
        renderSuggestion={renderSuggestion}
        onSuggestionSelected={onSuggestionSelected}
        shouldRenderSuggestions={value => true}
        focusInputOnSuggestionClick={false}
        inputProps={inputProps}
      />
    </Wrapper>
  );
};

export default Searchbox;

const Wrapper = styled.div`
  padding: 10px;

  .react-autosuggest {
    &__container {
      width: 100%;
      margin: 0 auto;
      position: relative;
      padding-right: 20px;
      
      @media (min-width: 1024px) {
        width: 300px;
      }
    }

    &__suggestions-container {
      position: absolute;
      width: 100%;
      box-shadow: 0 6px 7px 0 rgba(0, 0, 0, 0.1);
      max-height: 200px;
      overflow: auto;
      background-color: #ffffff;
    }
  }

  ul {
    padding: 0;
    margin: 0;
    list-style: none;

    li {
      padding: 10px;
      border-top: 1px solid #e5e5e5;
      cursor: pointer;
    }
  }
`;

const Suggestion = styled.div``;
