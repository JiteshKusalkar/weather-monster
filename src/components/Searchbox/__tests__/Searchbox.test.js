import { shallow } from 'enzyme';
import React from 'react';
import Autosuggest from 'react-autosuggest';
import unroll from 'unroll';

import Searchbox from '../';

const validProps = {
  id: 'city-search',
  onSuggestionSelect: jest.fn()
};

const emptyProps = {
  id: '',
  onSuggestionSelect: () => null
};

const shallowInstance = props => shallow(<Searchbox {...props} />);
unroll.use(it);

describe('Snapshot: Searchbox', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallowInstance(emptyProps);
  });

  unroll(
    'should match the snapshot for #propType',
    (done, args) => {
      wrapper.setProps(args.value);
      expect(wrapper).toMatchSnapshot();
      done();
    },
    [
      ['propType', 'value'],
      ['valid props', validProps],
      ['empty props', emptyProps]
    ]
  );
});

describe('Unit: Searchbox', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallowInstance(validProps);
  });

  it('should delete city on click of weather card', () => {
    const evt = {};
    const suggestion = {
      city: {
        id: 147059,
        name: 'Telmankend',
        findname: 'TELMANKEND',
        country: 'AZ',
        coord: { lon: 48.399021, lat: 39.87867 },
        zoom: 10
      }
    };
    wrapper.find(Autosuggest).simulate('suggestionSelected', evt, { suggestion });

    expect(validProps.onSuggestionSelect).toHaveBeenCalledWith(suggestion);
  });
});
