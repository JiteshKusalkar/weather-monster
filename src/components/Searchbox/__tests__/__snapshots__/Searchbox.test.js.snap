// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot: Searchbox should match the snapshot for "empty props" 1`] = `ShallowWrapper {}`;

exports[`Snapshot: Searchbox should match the snapshot for "valid props" 1`] = `ShallowWrapper {}`;
